pkg/kubelet/kubelet.go

```diff
@@ -75,42 +80,71 @@ func NewMainKubelet(
 	maxContainerCount int,
 	sourcesReady SourcesReadyFn,
 	clusterDomain string,
-	clusterDNS net.IP) (*Kubelet, error) {
+	clusterDNS net.IP,
+	masterServiceNamespace string) (*Kubelet, error) {
 	if resyncInterval <= 0 {
 		return nil, fmt.Errorf("invalid sync frequency %d", resyncInterval)
 	}
 	if minimumGCAge <= 0 {
 		return nil, fmt.Errorf("invalid minimum GC age %d", minimumGCAge)
 	}
+
+	serviceStore := cache.NewStore()
+	cache.NewReflector(&serviceLW{kubeClient, masterServiceNamespace}, &api.Service{}, serviceStore).Run()
+	serviceLister := &cache.StoreToServiceLister{serviceStore}
+
 	return &Kubelet{
-		hostname:              hostname,
-		dockerClient:          dockerClient,
-		etcdClient:            etcdClient,
-		rootDirectory:         rootDirectory,
-		resyncInterval:        resyncInterval,
-		networkContainerImage: networkContainerImage,
-		podWorkers:            newPodWorkers(),
-		dockerIDToRef:         map[dockertools.DockerID]*api.ObjectReference{},
-		runner:                dockertools.NewDockerContainerCommandRunner(dockerClient),
-		httpClient:            &http.Client{},
-		pullQPS:               pullQPS,
-		pullBurst:             pullBurst,
-		minimumGCAge:          minimumGCAge,
-		maxContainerCount:     maxContainerCount,
-		sourcesReady:          sourcesReady,
-		clusterDomain:         clusterDomain,
-		clusterDNS:            clusterDNS,
+		hostname:               hostname,
+		dockerClient:           dockerClient,
+		etcdClient:             etcdClient,
+		kubeClient:             kubeClient,
+		rootDirectory:          rootDirectory,
+		resyncInterval:         resyncInterval,
+		networkContainerImage:  networkContainerImage,
+		podWorkers:             newPodWorkers(),
+		dockerIDToRef:          map[dockertools.DockerID]*api.ObjectReference{},
+		runner:                 dockertools.NewDockerContainerCommandRunner(dockerClient),
+		httpClient:             &http.Client{},
+		pullQPS:                pullQPS,
+		pullBurst:              pullBurst,
+		minimumGCAge:           minimumGCAge,
+		maxContainerCount:      maxContainerCount,
+		sourcesReady:           sourcesReady,
+		clusterDomain:          clusterDomain,
+		clusterDNS:             clusterDNS,
+		serviceLister:          serviceLister,
+		masterServiceNamespace: masterServiceNamespace,
 	}, nil
 }
 
+type serviceLW struct {
+	client    *client.Client
+	namespace string
+}
+
+// TODO: the List() and Watch() methods are prime examples of where
```

Comment:
Yeah, this is actually set up wrongly at this point; it should be watching in `NamespaceAll`